NewTsRebase/rebase.cc

```diff
@@ -0,0 +1,136 @@
+#include "nan.h"
+#include "include/mpegts_stream_walker.h"
+#include "include/ts_rebase_impl.h"
+#include "include/common.h"
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include "include/constants.h"
+#include "include/utils.h"
+
+
+// object keys
+#define CONTEXT_EXPECTED_DTS			"expectedDts"
+#define CONTEXT_CORRECTION				"correction"
+#define CONTEXT_TOTAL_FRAME_DURATIONS	"totalFrameDurations"
+#define CONTEXT_TOTAL_FRAME_COUNT		"totalFrameCount"
+#define TIMESTAMP_MASK (0x1ffffffff)	// 33 bits
+#define REBASE_THRESHOLD (22500)		// 1/4 sec
+
+using namespace v8;
+using namespace node;
+
+static const char kaltura_object_type[] = "KalturaSyncTimecode";
+
+long get_suggested_correction(ID3v2_struct* obj)
+{
+    long suggested_correction;
+    if (obj==NULL)
+    {
+        return 0    ;
+    }
+    while (1){
+        suggested_correction=90*obj->UnixTimeStamp-obj->PTS;
+        if  (!strcmp(obj->objectType,kaltura_object_type)){   //TODO: change to constant
+            suggested_correction=90*obj->UnixTimeStamp-obj->PTS; //normalize it such the units is second*90,000
+            if (*obj->objectType=='\0'){
+                puts("ERROR oject type is null");
+            }
+            return suggested_correction;
+        }
+        obj=obj->next;
+        if (obj == NULL)
+            break;
+    }
+    return suggested_correction;
+    
+}
+
+NAN_METHOD(RebaseTs)
+{
+    NanScope();
+    
+    // validate args
+    if (args.Length() < 2)
+    {
+        return NanThrowTypeError("Function requires 2 arguments");
+    }
+    
+    if (!args[0]->IsObject())
+    {
+        return NanThrowTypeError("Argument 1 must be an object");
+    }
+    
+    if (!args[1]->IsObject() || !Buffer::HasInstance(args[1]))
+    {
+        return NanThrowTypeError("Argument 2 must be a buffer");
+    }
+    
+    // parse the context
+    ts_rebase_context_t context;
+    Local<Value> curValue;
+    Local<Object> inputContext = args[0]->ToObject();
+    
+    curValue = inputContext->Get(NanNew<String>(CONTEXT_EXPECTED_DTS));
+    context.expected_dts = curValue->IsNumber() ? curValue->IntegerValue() : NO_TIMESTAMP;
+    
+    curValue = inputContext->Get(NanNew<String>(CONTEXT_CORRECTION));
+    context.correction = curValue->IsNumber() ? curValue->IntegerValue() : 0;
+    
+    curValue = inputContext->Get(NanNew<String>(CONTEXT_TOTAL_FRAME_DURATIONS));
+    context.total_frame_durations = curValue->IsNumber() ? curValue->IntegerValue() : 0;
+    
+    curValue = inputContext->Get(NanNew<String>(CONTEXT_TOTAL_FRAME_COUNT));
+    context.total_frame_count = curValue->IsNumber() ? curValue->IntegerValue() : 0;
+    
```

Comment:
add to context  the value of the get_suggested_correction  (which ID3 tag was selected, which PTS, etc..)